### PR TITLE
Ensures compiled metrics resolve output names correctly

### LINKED ISSUE #21675: Incoherent behaviour in Model.compile between loss & metrics for dict output Models
Hey!

I believe there is a bug in keras 3 related to metrics in `Model.compile`. At the very least there is a significant difference from keras 2. I have an example below about this below, which I tested with keras 2.15 & 3.11.3:

```python
import numpy as np
import keras

print('keras version:', keras.__version__)

X = np.random.rand(100, 32)
Y1 = np.random.rand(100, 10)
Y2 = np.random.rand(100, 10)

def create_model():
    x = keras.Input(shape=(32,))
    y1 = keras.layers.Dense(10)(x)
    y2 = keras.layers.Dense(10)(x)
    return keras.Model(inputs = x, outputs = {'a': y1, 'b': y2})

model = create_model()
model.compile(optimizer = 'adam', loss = {'a': 'bce', 'b': 'mse'})
model.train_on_batch(X, {'a': Y1, 'b': Y2})
print('Successful training not using metrics\n')

try:
    model = create_model()
    model.compile(optimizer = 'adam', loss = {'a': 'bce', 'b': 'mse'}, metrics = {'a': ['mae', 'acc'], 'b': 'mse'})
    model.train_on_batch(X, {'a': Y1, 'b': Y2})
    print('Successful training using metrics\n')  # passes in keras 2.15
except ValueError as ex:
    print('Unsuccessful training with metrics')  # fails in keras 3.11.3
    print(ex)

model = create_model()
model.compile(optimizer = 'adam', loss = {'a': 'bce', 'b': 'mse'}, loss_weights = {'a': 1., 'b': 0.5})
model.train_on_batch(X, {'a': Y1, 'b': Y2})
print('Successful training using loss_weights\n')

x = keras.Input(shape=(32,))
y1 = keras.layers.Dense(10, name = 'a')(x)
y2 = keras.layers.Dense(10, name = 'b')(x)
model = keras.Model(inputs = x, outputs = {'a': y1, 'b': y2})
model.compile(optimizer = 'adam', loss = {'a': 'bce', 'b': 'mse'}, metrics = {'a': ['mae', 'acc'], 'b': 'mse'})
model.train_on_batch(X, {'a': Y1, 'b': Y2})
print('Successful even with metrics')
```
As we see in the example, in keras 3 the names in the dict output of the `Model` are only assumed for the `loss` (and `loss_weights`), but not for `metrics`. This does not seem like intended behaviour.

If we give the layer names then it works in keras 3, as discussed #21259 but that is not a fix that works in all codebases.

I also created [this message](https://github.com/keras-team/keras/issues/18467#issuecomment-3288095533) about this.

I have not looked at the keras code yet, but plan to, and if I can, I'll create a PR.

### PR DESCRIPTION
**Pull request by JyotinderSingh**
## Issue
Fixes #21675

In Keras 3.x, compiling models with dict outputs and dict metrics can fail when `CompileMetrics` receives `output_names` that don't match the dict keys of `y_pred`. The builder prefers `output_names` (often internal op names like `['dense', 'dense_1']`) over actual dict keys (e.g., `{'a', 'b'}`), causing `ValueError ("key 'a' does not correspond to any model output"`)

This change ensures that `CompileMetrics.build` resolves output names by preferring dict keys when `y_pred` is a dict, and honors `self.output_names` only if its set matches the dict key set (to preserve stable ordering). It stores the resolved names `(_resolved_output_names`) and uses them consistently for flattening (`_flatten_y`) and metric name prefixing, preserving correct mapping of dict outputs.

## Testing
Added `test_dict_outputs_ignore_mismatched_output_names` validating that dict metrics build and run when `output_names` are op-style names that don't match user dict keys, and verifies metric naming (`a_*`/`b_*`) and correct values.

## Colab
[colab notebook demonstrating complete training loop](https://colab.research.google.com/gist/JyotinderSingh/ad3a5d489f015734450ac1f21c39967b/keras-3-dictionary-metrix-fix.ipynb)

### CODE REVIEW

File: keras/src/trainers/compile_utils.py [hunk 1]
```diff
@@ -316,9 +324,10 @@ def _build_metrics_set(
         return flat_metrics
 
     def _flatten_y(self, y):
-        if isinstance(y, dict) and self.output_names:
+        names = getattr(self, "_resolved_output_names", self.output_names)
```
[the remark this answers is not in the document]
Comment (JyotinderSingh) [reply keras/src/trainers/compile_utils.py]: Done.

**Review by hertschuh — COMMENTED** [review]
This looks good for that case of a single dict of outputs.

But outputs can be nested to any level. I don't think this will work with let's say nested dictionaries for outputs. Can you add a unit test for nested dicts and see what happens?

**Comment by hertschuh** [comment]
Here is what I tested:

```python
import numpy as np
import keras

print('keras version:', keras.__version__)

X = np.random.rand(100, 32)
Y1 = np.random.rand(100, 10)
Y2 = np.random.rand(100, 10)
Y3 = np.random.rand(100, 10)

def create_model():
    x = keras.Input(shape=(32,))
    y1 = keras.layers.Dense(10)(x)
    y2 = keras.layers.Dense(10)(x)
    y3 = keras.layers.Dense(10)(x)
    return keras.Model(inputs = x, outputs = {'a': y1, 'b': {'c': y2, 'd': y3}})

model = create_model()
model.compile(optimizer = 'adam', loss = {'a': 'bce', 'b': {'c': 'mse', 'd': 'mse'}})
model.train_on_batch(X, {'a': Y1, 'b': {'c': Y2, 'd': Y3}})
print('Successful training not using metrics\n')

try:
    model = create_model()
    model.compile(
        optimizer = 'adam',
        loss = {'a': 'bce', 'b': {'c': 'mse', 'd': 'mse'}},
        metrics = {'a': ['mae', 'acc'], 'b':{'c': 'mse', 'd': 'mse'}}
    )
    model.train_on_batch(X, {'a': Y1, 'b': {'c': Y2, 'd': Y3}})
    print('Successful training using metrics')
except ValueError as ex:
    print('Unsuccessful training with metrics')
    print(ex)

model = create_model()
model.compile(
    optimizer = 'adam',
    loss = {'a': 'bce', 'b': {'c': 'mse', 'd': 'mse'}},
    loss_weights = {'a': 1., 'b': {'c': 0.5, 'd': 0.4}}
)
model.train_on_batch(X, {'a': Y1, 'b': {'c': Y2, 'd': Y3}})
print('Successful training using loss_weights\n')
```

Nested losses work, but not nested metrics, you get this error:
```
All entries in the sublists of the `metrics` dict should be metric objects. At key 'b', found the following sublist with unknown types: [{'c': 'mse', 'd': 'mse'}]
```

But the issue is somewhere else, not what you changed. So let's treat that as a separate issue.
